### PR TITLE
KAFKA-7477: Improve Streams close timeout semantics

### PR DESCRIPTION
**Pull request by nizhikov**
Second part of [KIP-358](https://cwiki.apache.org/confluence/display/KAFKA/KIP-358%3A+Migrate+Streams+API+to+Duration+instead+of+long+ms+times).

This changes based on [previous PR discussion](https://github.com/apache/kafka/pull/5682#discussion_r221473451).

Default `close` timeout is `30 seconds`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by nizhikov** [comment]
Hello, @vvcephei.

Can you take a look?

**Comment by nizhikov** [comment]
Tests passed.

@mjsax Can you take a look?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 1]
```diff
@@ -219,11 +219,7 @@ private boolean waitOnState(final State targetState, final long waitMs) {
             long elapsedMs = 0L;
             while (state != targetState) {
                 if (waitMs == 0) {
-                    try {
-                        stateLock.wait();
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:223]: Why do we remove this? Would it be better to call it with `stateLock.wait(0L);` ? (or better merge with case below and update condition to `if (waitMs >= elapsedMs)` ?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 2]
```diff
@@ -126,7 +126,7 @@
 public class KafkaStreams {
 
     private static final String JMX_PREFIX = "kafka.streams";
-    private static final int DEFAULT_CLOSE_TIMEOUT = 0;
+    private static final int DEFAULT_CLOSE_TIMEOUT = 30;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Why this change?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 3]
```diff
@@ -824,7 +820,7 @@ public void close() {
      * threads to join.
      * A {@code timeout} of 0 means to wait forever.
      *
-     * @param timeout  how long to wait for the threads to shutdown
+     * @param timeout  how long to wait for the threads to shutdown. Can't be negative. If {@code timeout=0} just checking the state and return immediately.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:821]: We cannot change the semantics of exiting `close(long, TimeUnit)` -- this would be a backward incompatible change. We can only change the semantics for the new `close(Duration)` method.

We also should point out the different semantics in L830:
```
@deprecated Use {@link #close(Duration)} instead; note, that {@link #close(Duration)} has different semantics and does not block on zero, e.g., `Duration.ofMillis(0)`.

**Review by mjsax — COMMENTED** [review]
Thanks for the PR. We need to preserve existing semantics. I guess, the simplest way will be, to introduce a ` private synchronized boolean close(final long timeout, final TimeUnit timeUnit, boolean newSemantics)` and call it with `true`/`false` from new and old `close()`. The `newSemantics` flag can be passed into `waitOnState()` (or we duplicate the old one as `waitOnStateBlocking` that we only call for `newSemantics==false`?

If there is another more elegant way, we can do it differently, too. Just an idea.

**Comment by mjsax** [comment]
We should also add a test `KafkaStreamsTest` to test that the new method throws for negative values and does not block for zero (ie, please add two tests)

**Comment by nizhikov** [comment]
Hello, @mjsax 

Looks like a didn't understand initial task properly :)
Thank you for the clarification!

1. `waitOnState` reworked.
2. Tests added.
3. Current methods behaviour preserved.

Please, take a look.

**Comment by nizhikov** [comment]
Failure unrelated.

Retest this, please.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 4]
```diff
@@ -213,12 +213,15 @@ public boolean isValidTransition(final State newState) {
...
 
-    private boolean waitOnState(final State targetState, final long waitMs) {
+    private boolean waitOnState(final State targetState, final long waitMs, final boolean newSemantics) {
         final long begin = time.milliseconds();
         synchronized (stateLock) {
             long elapsedMs = 0L;
             while (state != targetState) {
                 if (waitMs == 0) {
+                    if (newSemantics)
+                        return false;
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Nit: every conditional needs to have braces (per the code style)

not nit: I find this logic a little difficult to follow. Contrary to what @mjsax suggested, wouldn't it be pretty straightforward to map the old semantics on to the new ones like this:
* negative numbers => 0
* 0 => Long.MAX_VALUE
* all other arguments stay the same
?

Then, the old close method could just transform its arguments and call the new method, with no need to have this "new semantics" flag and an early return in the middle of the loop.

Comment (nizhikov) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: @vvcephei I like your proposal. Thanks!
Changed PR according to it.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 5]
```diff
@@ -219,11 +219,7 @@ private boolean waitOnState(final State targetState, final long waitMs) {
             long elapsedMs = 0L;
             while (state != targetState) {
                 if (waitMs == 0) {
-                    try {
-                        stateLock.wait();
-                    } catch (final InterruptedException e) {
-                        // it is ok: just move on to the next iteration
-                    }
+                    return false;
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: This seems to be the same condition as the `else` block, and actually, given that `waitMs` can't be negative, I think the `else` block is unreachable. Perhaps we can consolidate the logic?

**Review by vvcephei — COMMENTED** [review]
Thanks for the update; this seems easier to follow to me. I had one more question...

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 3 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:821]: Seems this comment was not addressed yet.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 6]
```diff
@@ -833,7 +829,24 @@ public void close() {
...
+
+        return close(timeoutMs);
+    }
+
+    /**
+     * @param timeoutMs  how long to wait for the threads to shutdown.
+     * @return {@code true} if all threads were successfully stopped&mdash;{@code false} if the timeout was reached
+     * before all threads stopped
+     */
+    private synchronized boolean close(final long timeoutMs) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: both public methods are `synchronized` already -- don't think we need to `synchronize` this one.

Nit: we don't need JavaDocs for `private` methods

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 7]
```diff
@@ -833,7 +829,24 @@ public void close() {
...
+        log.debug("Stopping Streams client with timeoutMillis = {} ms.", timeoutMs);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: This can be confusion to users if they call the old API, as we would print modified `timeoutMS` value -- we need to move the into the corresponding public methods (for old API, we need to log before we modify the timeout). It might also be good, to update the log statement for the old API by saying that deprecated API call is made and code should be updated?)

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 8]
```diff
@@ -912,7 +925,12 @@ public void run() {
      */
     public synchronized boolean close(final Duration timeout) throws IllegalArgumentException {
         ApiUtils.validateMillisecondDuration(timeout, "timeout");
-        return close(timeout.toMillis(), TimeUnit.MILLISECONDS);
+
+        final long timeoutMs = timeout.toMillis();
+        if (timeoutMs < 0)
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: nit: add `{ }` to block

**Review by mjsax — COMMENTED** [review]
Thanks for updating the PR. Some more comments.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 3 above)
Comment (nizhikov) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:821]: Fixed.

**Comment by nizhikov** [comment]
@mjsax @vvcephei 

Thanks for the review.
I addressed all your comments.

**Review by mjsax — APPROVED** [review]
LGTM. Waiting for second +1 before merging.

**Comment by nizhikov** [comment]
Hello @guozhangwang @bbejeck 

I've got +1 from @mjsax and need one more :)
Can you take a look at my PR?

**Review by bbejeck — APPROVED** [review]
LGTM

**Comment by guozhangwang** [comment]
LGTM!

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `2.1` branch.
